**Index**

1. [Import libraries](#import-libraries)
2. [Import data](#import-data)
3. [Nomenclatures](#nomenclatures)
3. [Cleaning data](#cleaning-data)
4. [Joining data](#joining-data)
5. [Final data](#final-data)


 ### Import libraries

In [1]:
import pandas as pd
import numpy as np
import os

import warnings
warnings.filterwarnings("ignore")

from Toolbox_ML import *

# Configurar Pandas para mostrar todas las columnas y filas
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

### Import data

In [2]:
# Define the paths to the CSV files
data_folder = '../data_sample'
data_1_path = os.path.join(data_folder, 'heart_disease_uci.csv')
data_2_path = os.path.join(data_folder, 'heart.csv')

In [3]:
# Read the CSV files
data_1 = pd.read_csv(data_1_path)
data_2 = pd.read_csv(data_2_path)

In [4]:
# Display the first few rows of each dataframe to verify
data_1.head()

id  age     sex    dataset               cp  trestbps   chol    fbs  \
0   1   63    Male  Cleveland   typical angina     145.0  233.0   True   
1   2   67    Male  Cleveland     asymptomatic     160.0  286.0  False   
2   3   67    Male  Cleveland     asymptomatic     120.0  229.0  False   
3   4   37    Male  Cleveland      non-anginal     130.0  250.0  False   
4   5   41  Female  Cleveland  atypical angina     130.0  204.0  False   

          restecg  thalch  exang  oldpeak        slope   ca  \
0  lv hypertrophy   150.0  False      2.3  downsloping  0.0   
1  lv hypertrophy   108.0   True      1.5         flat  3.0   
2  lv hypertrophy   129.0   True      2.6         flat  2.0   
3          normal   187.0  False      3.5  downsloping  0.0   
4  lv hypertrophy   172.0  False      1.4    upsloping  0.0   

                thal  num  
0       fixed defect    0  
1             normal    2  
2  reversable defect    1  
3             normal    0  
4             normal    0

In [5]:
data_2.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

### Nomenclatures

In [6]:
data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 920 entries, 0 to 919
Data columns (total 16 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   id        920 non-null    int64  
 1   age       920 non-null    int64  
 2   sex       920 non-null    object 
 3   dataset   920 non-null    object 
 4   cp        920 non-null    object 
 5   trestbps  861 non-null    float64
 6   chol      890 non-null    float64
 7   fbs       830 non-null    object 
 8   restecg   918 non-null    object 
 9   thalch    865 non-null    float64
 10  exang     865 non-null    object 
 11  oldpeak   858 non-null    float64
 12  slope     611 non-null    object 
 13  ca        309 non-null    float64
 14  thal      434 non-null    object 
 15  num       920 non-null    int64  
dtypes: float64(5), int64(3), object(8)
memory usage: 115.1+ KB


Data_1's nomenclatures

|Column|Description|
|------|-----------|
|id|Unique id for each patient|
|Age|Age of the patient in years|
|sex|Male/Female|
|dataset|place of study|
|cp | chest pain type, typical angina, atypical angina, non-anginal, asymptomatic|
|trestbps|resting blood pressure (in mm Hg on admission to the hospital)|
|chol|serum cholesterol in mg/dl|
|fbs|if fasting blood sugar > 120 mg/dl: True/False|
|restecg|resting electrocardiographic results, Values: [normal, stt abnormality, lv hypertrophy]|
|thalch| maximum heart rate achieved|
|exang|exercise-induced angina (True/ False)|
|oldpeak| ST depression induced by exercise relative to rest|
|slope|the slope of the peak exercise ST segment, values: 'downsloping', 'flat', 'upsloping'|
|ca|number of major vessels (0-3) colored by fluoroscopy|
|thal| [normal; fixed defect; reversible defect]|
|num|the predicted attribute 0=no heart disease; 1,2,3,4 = stages of heart disease|



In [7]:
data_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             918 non-null    int64  
 1   Sex             918 non-null    object 
 2   ChestPainType   918 non-null    object 
 3   RestingBP       918 non-null    int64  
 4   Cholesterol     918 non-null    int64  
 5   FastingBS       918 non-null    int64  
 6   RestingECG      918 non-null    object 
 7   MaxHR           918 non-null    int64  
 8   ExerciseAngina  918 non-null    object 
 9   Oldpeak         918 non-null    float64
 10  ST_Slope        918 non-null    object 
 11  HeartDisease    918 non-null    int64  
dtypes: float64(1), int64(6), object(5)
memory usage: 86.2+ KB


Data_2's nomenclatures

|Column|Description|
|------|-----------|
|Age|age of the patient [years]|
|Sex|sex of the patient [M: Male, F: Female]|
|ChestPainType|chest pain type [TA: Typical Angina, ATA: Atypical Angina, NAP: Non-Anginal Pain, ASY: Asymptomatic]|
|RestingBP|resting blood pressure [mm Hg]|
|Cholesterol|serum cholesterol [mm/dl]|
|FastingBS|fasting blood sugar [1: if FastingBS > 120 mg/dl, 0: otherwise]|
|RestingECG|resting electrocardiographic results, Values:  [Normal: Normal, ST: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV), LVH: showing probable or definite left ventricular hypertrophy by Estes' criteria]|
|MaxHR| maximum heart rate achieved [Numeric value between 60 and 202]|
|ExerciseAngina| exercise-induced angina [Y: Yes, N: No]|
|Oldpeak|ST [Numeric value measured in depression]|
|ST_Slope| the slope of the peak exercise ST segment [Up: upsloping, Flat: flat, Down: downsloping]|
|HeartDisease|output class [1: heart disease, 0: Normal]|

### Cleaning data

Como los datos reciben distintos nombres en ambos datasets, renombrarés las columnas del data_1 par que coincida con los del data_2.
Además, estandarizaré los valores de las columnas categóricas para que también coincidan.

In [8]:
# Renaming columns in data_1 to match those in data_2
data_1.rename(columns={
    'age':'Age',
    'sex': 'Sex',
    'cp': 'ChestPainType',
    'trestbps': 'RestingBP',
    'chol': 'Cholesterol',
    'fbs': 'FastingBS',
    'restecg': 'RestingECG',
    'thalch': 'MaxHR',
    'exang': 'ExerciseAngina',
    'oldpeak': 'Oldpeak',
    'slope': 'ST_Slope',
    'num': 'HeartDisease'
},inplace=True)

In [9]:
print("Data_1 with renamed columns:")
data_1.head()

Data_1 with renamed columns:


id  Age     Sex    dataset    ChestPainType  RestingBP  Cholesterol  \
0   1   63    Male  Cleveland   typical angina      145.0        233.0   
1   2   67    Male  Cleveland     asymptomatic      160.0        286.0   
2   3   67    Male  Cleveland     asymptomatic      120.0        229.0   
3   4   37    Male  Cleveland      non-anginal      130.0        250.0   
4   5   41  Female  Cleveland  atypical angina      130.0        204.0   

  FastingBS      RestingECG  MaxHR ExerciseAngina  Oldpeak     ST_Slope   ca  \
0      True  lv hypertrophy  150.0          False      2.3  downsloping  0.0   
1     False  lv hypertrophy  108.0           True      1.5         flat  3.0   
2     False  lv hypertrophy  129.0           True      2.6         flat  2.0   
3     False          normal  187.0          False      3.5  downsloping  0.0   
4     False  lv hypertrophy  172.0          False      1.4    upsloping  0.0   

                thal  HeartDisease  
0       fixed defect             0  
1             normal             2  
2  reversable defect             1  
3             normal             0  
4             normal             0

In [10]:
# Standardizing values for 'Sex' column
data_1['Sex'] = data_1['Sex'].map({'Male': 'M', 'Female': 'F'})
data_2['Sex'] = data_2['Sex'].map({'M': 'M', 'F': 'F'})

In [11]:
# Standardizing values for 'ChestPainType' column
data_1['ChestPainType'] = data_1['ChestPainType'].map({
    'typical angina': 'TA', 
    'atypical angina': 'ATA', 
    'non-anginal': 'NAP', 
    'asymptomatic': 'ASY'})

In [12]:
data_2['ChestPainType'] = data_2['ChestPainType'].map({
    'TA': 'TA', 
    'ATA': 'ATA', 
    'NAP': 'NAP', 
    'ASY': 'ASY'})

In [13]:
# Standardizing values for 'FastingBS' column
data_1['FastingBS'] = data_1['FastingBS'].astype(bool)
data_2['FastingBS'] = data_2['FastingBS'].astype(bool) #1/0 to True/False

In [14]:
# Standardizing values for 'RestingECG' column
data_1['RestingECG'] = data_1['RestingECG'].map({
    'normal': 'Normal', 
    'stt abnormality': 'ST', 
    'lv hypertrophy': 'LVH'})

In [15]:
data_2['RestingECG'] = data_2['RestingECG'].map({
    'Normal': 'Normal', 
    'ST': 'ST', 
    'LVH': 'LVH'})

In [16]:
# Standardizing values for 'ST_Slope' column
data_1['ST_Slope'] = data_1['ST_Slope'].map({
    'downsloping': 'Down', 
    'flat': 'Flat', 
    'upsloping': 'Up'})

In [17]:
data_1['ST_Slope'] = data_1['ST_Slope'].map({
    'Down': 'Down', 
    'Flat': 'Flat', 
    'Up': 'Up'})

In [18]:
# Standardizing values for 'ExerciseAngina' column
data_1['ExerciseAngina'] = data_1['ExerciseAngina'].map({True: 'Y', False: 'N'})
data_2['ExerciseAngina'] = data_2['ExerciseAngina'].map({'Y': 'Y', 'N': 'N'})

In [19]:
# Standardizing values for 'HeartDisease' column
data_1['HeartDisease'] = data_1['HeartDisease'].apply(lambda x: 1 if x > 0 else 0)
data_2['HeartDisease'] = data_2['HeartDisease'].astype(int)

In [20]:
print("Data_1 standarized:")
data_1.head()

Data_1 standarized:


id  Age Sex    dataset ChestPainType  RestingBP  Cholesterol  FastingBS  \
0   1   63   M  Cleveland            TA      145.0        233.0       True   
1   2   67   M  Cleveland           ASY      160.0        286.0      False   
2   3   67   M  Cleveland           ASY      120.0        229.0      False   
3   4   37   M  Cleveland           NAP      130.0        250.0      False   
4   5   41   F  Cleveland           ATA      130.0        204.0      False   

  RestingECG  MaxHR ExerciseAngina  Oldpeak ST_Slope   ca               thal  \
0        LVH  150.0              N      2.3     Down  0.0       fixed defect   
1        LVH  108.0              Y      1.5     Flat  3.0             normal   
2        LVH  129.0              Y      2.6     Flat  2.0  reversable defect   
3     Normal  187.0              N      3.5     Down  0.0             normal   
4        LVH  172.0              N      1.4       Up  0.0             normal   

   HeartDisease  
0             0  
1             1  
2             1  
3             0  
4             0

In [21]:
data_1.ST_Slope.unique()

array(['Down', 'Flat', 'Up', nan], dtype=object)

In [22]:
data_2.ST_Slope.unique()

array(['Up', 'Flat', 'Down'], dtype=object)

### Deleting nulls and merging datasets

In [23]:
describe_df(data_1)

id    Age     Sex dataset ChestPainType RestingBP  \
Tipos           int64  int64  object  object        object   float64   
% Faltante        0.0    0.0     0.0     0.0           0.0      6.41   
Valores Únicos    920     50       2       4             4        61   
% Cardinalidad  100.0   5.43    0.22    0.43          0.43      6.63   

               Cholesterol FastingBS RestingECG    MaxHR ExerciseAngina  \
Tipos              float64      bool     object  float64         object   
% Faltante            3.26       0.0      19.67     5.98           5.98   
Valores Únicos         217         2          2      119              2   
% Cardinalidad       23.59      0.22       0.22    12.93           0.22   

                Oldpeak ST_Slope       ca    thal HeartDisease  
Tipos           float64   object  float64  object        int64  
% Faltante         6.74    33.59    66.41   52.83          0.0  
Valores Únicos       53        3        4       3            2  
% Cardinalidad     5.76     0.33     0.43    0.33         0.22

Como ya habíamos visto anteriormente, el data_1 tiene muchos nulos en varias columnas, para evitar problemas y para no crear datos falsos que puedan falsear los resultados finales, eliminaremos las filas que contengan valores nulos y también nos desharemos de las columnas que no se repitan en el data_2 antes de unirlos.

In [24]:
# Deleting rows with null values in data_1
data_1.dropna(inplace=True)

In [25]:
# Verification of deletion of nulls
print("\nData_1's info after deletion of nulls:")
describe_df(data_1)


Data_1's info after deletion of nulls:


id    Age     Sex dataset ChestPainType RestingBP  \
Tipos           int64  int64  object  object        object   float64   
% Faltante        0.0    0.0     0.0     0.0           0.0       0.0   
Valores Únicos    295     40       2       3             4        49   
% Cardinalidad  100.0  13.56    0.68    1.02          1.36     16.61   

               Cholesterol FastingBS RestingECG    MaxHR ExerciseAngina  \
Tipos              float64      bool     object  float64         object   
% Faltante             0.0       0.0        0.0      0.0            0.0   
Valores Únicos         152         2          2       91              2   
% Cardinalidad       51.53      0.68       0.68    30.85           0.68   

                Oldpeak ST_Slope       ca    thal HeartDisease  
Tipos           float64   object  float64  object        int64  
% Faltante          0.0      0.0      0.0     0.0          0.0  
Valores Únicos       39        3        4       3            2  
% Cardinalidad    13.22     1.02     1.36    1.02         0.68

In [26]:
# Deleting data_1's columns that aren't on data_2
print(data_1.columns)
print("---")
print(data_2.columns)

Index(['id', 'Age', 'Sex', 'dataset', 'ChestPainType', 'RestingBP',
       'Cholesterol', 'FastingBS', 'RestingECG', 'MaxHR', 'ExerciseAngina',
       'Oldpeak', 'ST_Slope', 'ca', 'thal', 'HeartDisease'],
      dtype='object')
---
Index(['Age', 'Sex', 'ChestPainType', 'RestingBP', 'Cholesterol', 'FastingBS',
       'RestingECG', 'MaxHR', 'ExerciseAngina', 'Oldpeak', 'ST_Slope',
       'HeartDisease'],
      dtype='object')


In [27]:
common_columns = list(set(data_1.columns).intersection(set(data_2.columns)))
data_1 = data_1[common_columns]
data_2 = data_2[common_columns]

In [28]:
print(data_1.columns)
print("---")
print(data_2.columns)

Index(['FastingBS', 'HeartDisease', 'ChestPainType', 'MaxHR', 'ExerciseAngina',
       'ST_Slope', 'RestingECG', 'Sex', 'Cholesterol', 'Oldpeak', 'RestingBP',
       'Age'],
      dtype='object')
---
Index(['FastingBS', 'HeartDisease', 'ChestPainType', 'MaxHR', 'ExerciseAngina',
       'ST_Slope', 'RestingECG', 'Sex', 'Cholesterol', 'Oldpeak', 'RestingBP',
       'Age'],
      dtype='object')


In [29]:
# Merging datasets
merged_data = pd.concat([data_1, data_2], ignore_index=True)

In [32]:
# Ordenar las columnas alfabéticamente
merged_data = merged_data[sorted(merged_data.columns)]

### Export cleaned and merged datas to a CSV's file

In [33]:
output_path = os.path.join(data_folder, 'data_final.csv')
merged_data.to_csv(output_path, index=False)

In [34]:
# View of changes
print("\nMerged data:")
merged_data.head()


Merged data:


Age ChestPainType  Cholesterol ExerciseAngina  FastingBS  HeartDisease  \
0   63            TA        233.0              N       True             0   
1   67           ASY        286.0              Y      False             1   
2   67           ASY        229.0              Y      False             1   
3   37           NAP        250.0              N      False             0   
4   41           ATA        204.0              N      False             0   

   MaxHR  Oldpeak  RestingBP RestingECG ST_Slope Sex  
0  150.0      2.3      145.0        LVH     Down   M  
1  108.0      1.5      160.0        LVH     Flat   M  
2  129.0      2.6      120.0        LVH     Flat   M  
3  187.0      3.5      130.0     Normal     Down   M  
4  172.0      1.4      130.0        LVH       Up   F